### PR TITLE
Vendoring libnetwork @c8ce8c7

### LINKED ISSUE #23980: Unclear error message when overlay network is created with invalid parameters
While trying out overlay network and starting containers on it, I got following error:

```
docker: Error response from daemon: invalid bit range [4, 4].
```
I think that the reason was invalid parameters used in network creation (managed to get things working by removing those), but the error message is somewhat unclear.

**Output of `docker version`:**

```
Docker version 1.11.2, build b9f10c9
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 1
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: host overlay bridge null
Kernel Version: 4.4.0-24-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.953 GiB
Name: infra-front
ID: QAPB:EYGZ:K3X2:QKM2:IHYQ:EO7U:VPWM:2XJC:E7KS:W6VY:NFBD:HQ7U
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Cluster store: consul://192.168.80.10:8500
Cluster advertise: 192.168.80.10:2370
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Virtualbox

**Steps to reproduce the issue:**
1. docker network create -d overlay --subnet=192.168.50.0/24 --gateway=192.168.50.1 --ip-range=192.168.50.4/32 test
2. docker run --rm -ti --net test alpine sh

**Describe the results you received:**
Docker reports error,
```
docker: Error response from daemon: invalid bit range [4, 4].
```
**Describe the results you expected:**
More information about what is wrong, and how to fix this.

**Additional information you deem important (e.g. issue happens only occasionally):**

### LINKED ISSUE #26039: Specifying "--dns-opt ndots:X" should override the default "ndots:0"
**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:15:28 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:15:28 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 1
Server Version: 1.12.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host overlay bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.15-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.954 GiB
Name: moby
ID: UI36:ZFB5:JVVI:6TR5:YI4Z:7ZAG:DI5M:UBOM:DCJW:OFXX:O4E5:BP5U
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 18
 Goroutines: 29
 System Time: 2016-08-26T09:54:35.208231913Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Username: fermayo
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Using Docker for Mac

**Steps to reproduce the issue:**
1. `docker network create test`
2. `docker run --net test --dns-opt ndots:1 alpine cat /etc/resolv.conf`

**Describe the results you received:**

```
search local
nameserver 127.0.0.11
options ndots:1 ndots:0
```

**Describe the results you expected:**

```
search local
nameserver 127.0.0.11
options ndots:1
```

**Additional information you deem important (e.g. issue happens only occasionally):**

### PR DESCRIPTION
**Pull request by aboch**
Fixes #26039
Fixes #23980 
Brings in changes needed to address #11518

### CODE REVIEW

**Comment by thaJeztah** [comment]
@aboch are these fixes for 1.13, or a possible 1.12.2?

File: vendor/src/github.com/docker/libnetwork/controller.go [hunk 1]
```diff
@@ -614,6 +617,15 @@ func (c *controller) RegisterDriver(networkType string, driver driverapi.Driver,
 // NewNetwork creates a new network of the specified network type. The options
 // are network specific and modeled in a generic way.
 func (c *controller) NewNetwork(networkType, name string, id string, options ...NetworkOption) (Network, error) {
+	if id != "" {
+		c.networkLocker.Lock(id)
+		defer c.networkLocker.Unlock(id)
```
Comment (cpuguy83) [inline vendor/src/github.com/docker/libnetwork/controller.go]: I know this is from the vendored code, but does this need to be locked for the duration of the function or just the if block?

**Comment by aboch** [comment]
@thaJeztah 1.12.2

**Comment by aaronlehmann** [comment]
Will this version of libnetwork be compatible with the version of swarmkit we are using on the 1.12 branch? We aren't going to ship swarmkit master in 1.12.2.

**Comment by aboch** [comment]
@aaronlehmann confirmed with @mrjana there should not be issues in regard to that.

File: vendor/src/github.com/docker/libnetwork/controller.go (hunk 1 above)
Comment (aboch) [inline vendor/src/github.com/docker/libnetwork/controller.go]: @cpuguy83 Posting now a reply from this morning, just realized I had to submit the review...

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by tonistiigi — APPROVED** [review]
LGTM

**Pull request merged**